### PR TITLE
core, eth, trie: bloom filter for trie node dedup during fast sync

### PR DESCRIPTION
**Pull request by karalabe**
State sync (part of fast sync) works by taking the root of a state trie, downloading it from remote peers, deduplicating all children based on the local database and recursing any missing subtries. The deduplication simply checks the local database if the node is already known and omits downloading it if so. Unfortunately this check is exceedingly expensive as the database grows.

---

This PR adds a bloom filter to the state sync, so that instead of checking each and every node whether we already have it locally on disk or not, we first consult a gigantic bloom filter. If the node being deduplicated is not in the bloom, it surely is not on disk either, so we can save an expensive database lookup. If the bloom says it "might" be present, we must verify.

The bloom however can only return correct answers (i.e. a trie node surely was not yet downloaded) if we ensure that all trie nodes on disk have been injected into the filter too. This is the case for a fresh node with no database (i.e. no nodes whatsoever), but it does not hold true if a node is restarted mid-sync. For those instances, the PR starts by scanning the local database and injecting any previously downloaded trie node hashes into the bloom. As long as the initialization is running, the bloom always returns "maybe", delegating fast sync to reach down into the database instead of relying on the bloom.

---

The code is not ideal. We need to clean it up a bit, possibly thread safety wise too. We might also consider reimplementing the bloom instead of pulling in the library, not a fan. The PR is mostly up for benchmarking, discussions, reviews (note, it's not **bad**, just not final).

### CODE REVIEW

File: eth/downloader/statesync.go [hunk 1]
```diff
@@ -59,6 +59,13 @@ type stateSyncStats struct {
 
 // syncState starts downloading state with the given root hash.
 func (d *Downloader) syncState(root common.Hash) *stateSync {
+	// The downloader was requested to start a state sync. If the state sync bloom
+	// filter was not yet initialized, create it now. This lazy creation ensures we
+	// only allocate if if really really really needed.
```
Comment (holiman) [inline eth/downloader/statesync.go]:
```suggestion
	// only allocate if really really really needed.
```
`3` x `really` looks intentional, but `2` x `if` looks like a bug

File: trie/sync.go [hunk 2]
```diff
@@ -292,8 +295,11 @@ func (s *Sync) children(req *request, object node) ([]*request, error) {
 			if _, ok := s.membatch.batch[hash]; ok {
 				continue
 			}
-			if ok, _ := s.database.Has(node); ok {
-				continue
+			if s.bloom.Contains(node) {
```
Comment (holiman) [inline trie/sync.go]: There are two more location where (I think) the bloom should be used. Around line 101 in `AddSubTrie` and line 139 in `AddRawEntry`

File: trie/sync.go [hunk 3]
```diff
@@ -292,8 +295,11 @@ func (s *Sync) children(req *request, object node) ([]*request, error) {
...
+				// Bloom filter says this might be a duplicate, double check
+				if ok, _ := s.database.Has(node); ok {
+					continue
+				}
```
Comment (holiman) [inline trie/sync.go]: At this point, I'd like an else-block we note that the bloom delivered a FP, with a metric. That way, we can track bloom saturation and see how it degrades.

File: trie/sync_bloom.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	it := database.NewIterator()
+	defer it.Release()
+
+	for it.Next() && atomic.LoadUint32(&b.closed) == 0 {
+		if key := it.Key(); len(key) == common.HashLength {
+			b.bloom.Add(syncBloomHasher(key))
+			bloomLoadMeter.Mark(1)
+		}
+	}
+	log.Info("Initialized fast sync bloom", "items", b.bloom.N(), "errorrate", b.errorRate())
```
Comment (holiman) [inline trie/sync_bloom.go]: Might be interesting to also output the time that the iteration took

File: trie/sync_bloom.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+	// Mark the bloom filter inited and return
+	defer atomic.StoreUint32(&b.inited, 1)
```
Comment (holiman) [inline trie/sync_bloom.go]:
```suggestion
	defer atomic.StoreUint32(&b.inited, 1)
```
it's the last statement... ?

**Comment by holiman** [comment]
On my machine ( a VM with NAT)
```
[user@work go-ethereum]$ build/bin/geth
INFO [04-23|12:55:51.680] Bumping default cache on mainnet         provided=1024 updated=4096
...
INFO [04-23|12:56:41.861] Block synchronisation started 
INFO [04-23|12:56:47.231] Allocated fast sync bloom                size=1294.00MiB
...
INFO [04-23|12:57:07.142] Imported new block headers               count=0    elapsed=6.903ms   number=1232267 hash=1edd29…4ead71 age=3y1mo1w ignored=192
INFO [04-23|12:57:07.206] Initialized fast sync bloom              items=2431935 errorrate=0.000 time=19.974763615s
...

Basically, it currently starts iterating the state lazily once we get some state from a peer. So on my machine it's `50s` before I get a good peer, and only then the bloom iteration starts, and took `20s`. Wouldn't it be better to start the bloom init as soon as we determine that we're going to do a fast-sync?

File: trie/sync.go (hunk 2 above)
Comment (karalabe) [reply trie/sync.go]: Ah yeah, I was looking a those a while back too. Indeed could help a bit.

File: trie/sync.go (hunk 3 above)
Comment (holiman) [reply trie/sync.go]: Ah, I see, you're calculating it theoretically instead. Yes, I guess that works fine too

File: trie/sync_bloom.go (hunk 5 above)
Comment (karalabe) [reply trie/sync_bloom.go]: Ah yeah, I moved it from the front, copy paste error. thx

File: trie/sync_bloom.go (hunk 4 above)
Comment (karalabe) [reply trie/sync_bloom.go]: done

File: trie/sync.go (hunk 3 above)
Comment (karalabe) [reply trie/sync.go]: done

**Comment by holiman** [comment]
It finished fast sync in roughly 6 hours, bloom error rate reaching `100`, which is `1%` error rate. Db read `1.8T`, It was fast enough that it even suffered from a leveldb-imposed write delay, which we usually does not get after Gary's fixes.

File: trie/sync.go [hunk 6]
```diff
@@ -219,8 +230,9 @@ func (s *Sync) Commit(dbw ethdb.Writer) (int, error) {
 		if err := dbw.Put(key[:], s.membatch.batch[key]); err != nil {
 			return i, err
 		}
+		s.bloom.Add(key[:])
 	}
-	written := len(s.membatch.order)
+	written := len(s.membatch.order) // TODO(karalabe): could an order change improve write performance?
```
Comment (Matthalp) [inline trie/sync.go]: FWIW I wondered if the `order` field in `Batch` was still adding value in the codebase. But that's probably because I haven't been contributing for that long.

File: trie/sync.go [hunk 7]
```diff
@@ -292,8 +304,13 @@ func (s *Sync) children(req *request, object node) ([]*request, error) {
 			if _, ok := s.membatch.batch[hash]; ok {
 				continue
 			}
-			if ok, _ := s.database.Has(node); ok {
-				continue
+			if s.bloom.Contains(node) {
```
Comment (Matthalp) [inline trie/sync.go]: Similar comment about common code above.

File: trie/sync_bloom.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+type SyncBloom struct {
+	bloom  *bloomfilter.Filter
+	inited uint32
+	closer sync.Once
+	closed uint32
+	pend   sync.WaitGroup
+}
+
+// NewSyncBloom creates a new bloom filter of the given size (in megabytes) and
+// initializes it from the database. The bloom is hard coded to use 3 filters.
```
Comment (Matthalp) [inline trie/sync_bloom.go]: Consider adding some more information about why these bloom filter parameters are used. For example:
1. What is the desired false positive rate?
2. What is the expected number of entries to handle?

File: trie/sync_bloom.go [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+func NewSyncBloom(memory uint64, database ethdb.Iteratee) *SyncBloom {
```
Comment (Matthalp) [inline trie/sync_bloom.go]: Consider renaming `memory` to `size` (or even better `sizeMB`).

File: trie/sync_bloom.go [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	// Create the bloom filter to track known trie nodes
+	bloom, err := bloomfilter.New(memory*1024*1024*8, 3)
```
Comment (Matthalp) [inline trie/sync_bloom.go]: Consider making this a constant (`bitsInMB` or `bitsInMegabyte`).

File: trie/sync_bloom.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+	if err != nil {
+		panic(fmt.Sprintf("failed to create bloom: %v", err)) // Can't happen, here for sanity
+	}
+	log.Info("Allocated fast sync bloom", "size", common.StorageSize(memory*1024*1024))
```
Comment (Matthalp) [inline trie/sync_bloom.go]: I probably like constants too much. I would have made this `megabyte`.

File: trie/sync_bloom.go [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+
+	// Assemble the fast sync bloom and init it from previous sessions
+	b := &SyncBloom{
+		bloom: bloom,
+	}
+	b.pend.Add(2)
```
Comment (Matthalp) [inline trie/sync_bloom.go]: I'm a pretty big fan of the [`errgroup`](https://godoc.org/golang.org/x/sync/errgroup) library (especially with the `ctx.Done()` pattern). It might be worth checking out sometime.

File: trie/sync_bloom.go [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+// meter periodically recalculates the false positive error rate of the bloom
+// filter and reports it in a metric.
+func (b *SyncBloom) meter() {
+	for {
+		// Report the current error ration. No floats, lame, scale it up.
+		bloomErrorGauge.Update(int64(b.errorRate() * 100000))
+
+		// Wait one second, but check termination more frequently
+		for i := 0; i < 10; i++ {
+			if atomic.LoadUint32(&b.closed) == 1 {
```
Comment (Matthalp) [inline trie/sync_bloom.go]: FWIW This is a place where the `errgroup` `ctx.Done()` pattern in a `select` block would work out nicely.

File: trie/sync_bloom.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+// Add inserts a new trie node hash into the bloom filter.
+func (b *SyncBloom) Add(hash []byte) {
+	if atomic.LoadUint32(&b.closed) == 1 {
+		return
+	}
+	b.bloom.Add(syncBloomHasher(hash))
+	bloomAddMeter.Mark(1)
+}
+
+// Contains tests if the bloom filter contains  the given hash:
```
Comment (Matthalp) [inline trie/sync_bloom.go]: There is a double space: "contains  the given".

File: trie/sync_bloom.go [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+//   - false: the bloom definitely does not contain hash
+//   - true:  the bloom maybe contains hash
+//
+// While the bloom is being initialized, any query will return true.
+func (b *SyncBloom) Contains(hash []byte) bool {
```
Comment (Matthalp) [inline trie/sync_bloom.go]: The disconnect between the godoc and current method name suggests that `MightContain` could be a better name.

File: trie/sync_bloom.go [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+		bloomMissMeter.Mark(1)
+	}
+	return maybe
+}
+
+// errorRate calculates the probability of a random containment test returning a
+// false positive.
+//
+// We're calculating it ourselves because the bloom library we used missed a
+// paranthesis in the formula and calculates it wrong. And it's discontinued...
```
Comment (Matthalp) [inline trie/sync_bloom.go]: :(

Comment (Matthalp) [inline trie/sync_bloom.go]: s/paranthesis/parentheses/

**Comment by Matthalp** [comment]
I think this is a great idea to pursue. I do wonder how future-proof the bloom filter and its parameters are as the state size grows. Has any analysis been done on that?

**Comment by holiman** [comment]
Fast sync finished, 

- `7h 45m` vs  `11h 35m`
- `1.84TB` db read vs `3.76TB` read
- `1.89TB` db write vs `2.76 TB` write

File: trie/sync.go [hunk 17]
```diff
@@ -134,8 +140,13 @@ func (s *Sync) AddRawEntry(hash common.Hash, depth int, parent common.Hash) {
 	if _, ok := s.membatch.batch[hash]; ok {
 		return
 	}
-	if ok, _ := s.database.Has(hash.Bytes()); ok {
-		return
+	if s.bloom.Contains(hash[:]) {
```
[the remark this answers is not in the document]
Comment (karalabe) [reply trie/sync.go]: Meh, I'm not a fan of making every tiny op a method. Unless the code is really complex, I think inlining some minor things keeps the mental overhead of reading it lower. 

There's also a slight difference. This code does a `return`, code a bit down does `continue`, so you still need to check the result and do something accordingly. You'd just replace 5 LOC with a method declaration + 3 LOC to call it.

File: trie/sync_bloom.go (hunk 8 above)
Comment (karalabe) [reply trie/sync_bloom.go]: There's no desired rate or expected entries. All we know is that we have a fixed amount of memory we can use. We max it out and hope for the best. I can give an example in the comment, but there's no way to tune this really, since we don't know which network we run against or how many entries there will be.

File: trie/sync_bloom.go (hunk 10 above)
Comment (karalabe) [reply trie/sync_bloom.go]: :P Making this a constant is just mental burden, since you need to look up one extra thing in the source code to understand it. I think it's fairly clear here what it means.

File: trie/sync_bloom.go (hunk 11 above)
Comment (karalabe) [reply trie/sync_bloom.go]: :P Making this a constant is just mental burden, since you need to look up one extra thing in the source code to understand it. I think it's fairly clear here what it means.

File: trie/sync_bloom.go (hunk 15 above)
Comment (karalabe) [reply trie/sync_bloom.go]: Well, the bloom *does* contain this hash. The bloom **containing** the hash means the **database** *might* contain it. If this were a method on the database, I'd agree, but since it's a method on the bloom, it's correct.

File: trie/sync_bloom.go (hunk 13 above)
Comment (karalabe) [reply trie/sync_bloom.go]: Yeah, this was me being lazy. I'll swap it ouf for channels, just wanted to prep a PR for a benchmark.

File: eth/downloader/downloader.go [hunk 18]
```diff
@@ -104,7 +105,10 @@ type Downloader struct {
 	genesis    uint64   // Genesis block number to limit sync to (e.g. light client CHT)
 	queue      *queue   // Scheduler for selecting the hashes to download
 	peers      *peerSet // Set of active peers from which download can proceed
-	stateDB    ethdb.Database
+
+	stateDatabase  ethdb.Database  // Database to state sync into (and deduplicate via)
```
[the remark this answers is not in the document]
Comment (karalabe) [reply eth/downloader/downloader.go]: Reverted.
